### PR TITLE
Beginning C++23: Chapter 12

### PR DESCRIPTION
**Pull request by Peter-VanWeert**

### CODE REVIEW

File: Exercises/Modules/Chapter 12/Soln12_05/Soln12_05.cpp [hunk 1]
```diff
@@ -1,42 +1,30 @@
...
-  
-  Integer::printCount();   // 4 objects
-  
-  showIntegerRef(array[0]);
-  showIntegerVal(array[1]);
-  
-  Integer::printCount();   // 4 objects again
+  std::println("Result of comparing i and j is {}", compare(i, j));
+  std::println("Result of comparing k and j is {}", compare(k, j));
 }
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_05/Soln12_05.cpp:30]: Add a call to nonFriendCompare()?

File: Exercises/Modules/Chapter 12/Soln12_06/Integer.cppm [hunk 2]
```diff
@@ -0,0 +1,22 @@
+export module integer;
+
+export class Integer
+{
+public:
+  Integer(int value = 0);                  // Contructor with given value
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_06/Integer.cppm:6]: typo -> Constructor

File: Exercises/Modules/Chapter 12/Soln12_07/Soln12_07.cpp [hunk 3]
```diff
@@ -1,39 +1,66 @@
...
-   To show reverse iteration, we've modified findSmallestBox() to iterate in reverse order
- */
+/* 
+ Turns out moving the getFirstBox() and getLastBox() functions 
+ to the nested iterator class is easy. We used the same names for
+ the member variables, so it is just a matter of changing "Truckload::"
+ into "Truckload::Iterator::".
+ 
+ Makes sense: the logic for iterating the linked lists has not changed.
+ The only difference is that now there are m_current and m_head pointers
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_07/Soln12_07.cpp:13]: m_current -> m_next

File: Exercises/Modules/Chapter 12/Soln12_08/Truckload.cppm [hunk 4]
```diff
@@ -47,7 +43,6 @@ public:
   SharedBox getLastBox();     // Get the first Box
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_08/Truckload.cppm:43]: comment is wrong.

File: Exercises/Modules/Chapter 12/Soln12_09/Truckload.cppm [hunk 5]
```diff
@@ -39,13 +42,21 @@ private:
 class Truckload::Iterator
 {
 public:
-  SharedBox getFirstBox();  // Get the first Box
-  SharedBox getNextBox();   // Get the next Box
+  SharedBox getFirstBox();    // Get the first Box
+  SharedBox getLastBox();     // Get the first Box
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_09/Truckload.cppm:46]: wrong comment.

File: Exercises/Modules/Chapter 12/Soln12_09/Truckload.cpp [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  }
+
+  return false;     // Return false: boxToRemove was not found
+}
+
+bool Truckload::removeBox(Iterator& iter)
+{
+  if (iter.m_current)
+  {
+    auto* next = iter.m_current->m_next;
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_09/Truckload.cpp:145]: uniform init?

File: Exercises/Modules/Chapter 12/Soln12_09/Truckload.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+  return m_current? m_current->m_box : nullptr;   // Return its box (or nullptr...).
+}
+
+SharedBox Truckload::Iterator::getPreviousBox()
+{
+  if (m_current)                                 
+    m_current = m_current->m_previous;           // Move to the previous package
+
+  return m_current ? m_current->m_box : nullptr; // Return its box (or nullptr...).
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_09/Truckload.cpp:104]: return getCurrentBox();
?

File: Exercises/Modules/Chapter 12/Soln12_09/Truckload.cpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  m_current = m_tail;
+  return getCurrentBox();
+}
+
+SharedBox Truckload::Iterator::getNextBox()
+{
+  if (m_current)                                 
+    m_current = m_current->m_next;                // Move to the next package
+
+  return m_current? m_current->m_box : nullptr;   // Return its box (or nullptr...).
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_09/Truckload.cpp:96]: return getCurrentBox();
?

File: Examples/Modules/Chapter 12/Ex12_19/Ex12_19.cpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  std::vector<SharedBox> boxes;        // Array of Box objects
+
+  for (std::size_t i {} ; i < nBoxes ; ++i)
+    boxes.push_back(randomSharedBox());
+
+  Truckload load2{boxes};
+  std::println("\nThe second list:");
+  load2.printBoxes();
+
+  auto smallestBox = load2.getFirstBox();
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 12/Ex12_19/Ex12_19.cpp:49]: uniform init?

File: Examples/Modules/Chapter 12/Ex12_19/Ex12_19.cpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  for (auto box = load2.getNextBox(); box; box = load2.getNextBox())
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 12/Ex12_19/Ex12_19.cpp:50]: uniform init?

File: Examples/Modules/Chapter 12/Ex12_19/Truckload.cpp [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  m_next = m_head;      // nullptr only for an empty truckload
+  return getNextBox();
+}
+
+SharedBox Truckload::getNextBox()
+{
+  if (!m_next)          // If there's no next...
+    return nullptr;     // ...return nullptr
+                                    
+  SharedBox result = m_next->m_box; // Extract the box to return
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 12/Ex12_19/Truckload.cpp:51]: uniform init?

File: Exercises/Modules/Chapter 12/Soln12_07/Truckload.cpp [hunk 12]
```diff
@@ -46,111 +44,74 @@ Truckload::~Truckload()
...
-  if (!m_current)                                 // If there's no current...
-    return getLastBox();                          // ...return the last Box
-
-  m_current = m_current->m_previous;              // Move to the next package
-
-  return m_current ? m_current->m_box : nullptr;  // Return its box (or nullptr...).
+  if (!m_next)          // If there's no next...
+    return nullptr;     // ...return nullptr
+                                    
+  SharedBox result = m_next->m_box; // Extract the box to return
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_07/Truckload.cpp:74]: uniform init?

File: Exercises/Modules/Chapter 12/Soln12_02/Integer.cppm [hunk 13]
```diff
@@ -25,7 +25,7 @@ public:
 //  int compare(Integer obj) const;        // Compare function with value parameter
   int compare(const Integer& obj) const;   // Compare function with reference parameter
 
-  void show() const;
+  void printValue() const;
```
[the remark this answers is not in the document]
Comment (Peter-VanWeert) [reply Exercises/Modules/Chapter 12/Soln12_02/Integer.cppm:28]: default argument value, even...

**Pull request merged**